lib/rubocop/cop/rspec/rails/persistence_called_outside_example.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      module Rails
+        # Checks for persistence calls outside example blocks.
+        #
+        # Prevents persistence calls outside examples which usually run
+        # with some sort of cleanup hook. Saving outside of example causes
+        # records to leak into other tests.
+        #
+        # @ example
+        #   # bad - records created outside example group
+        #   describe User do
+        #     User.create!
+        #   end
+        #
+        #   describe User do
+        #     user = User.new
+        #     user.save!
+        #   end
+        #
+        #   # good - records created inside an example group
+        #   describe User do
+        #     it do
+        #       User.create!
+        #     end
+        #   end
+        #
+        #   describe User do
+        #     it do
+        #       user = User.new
+        #       user.save!
+        #     end
+        #   end
+        class PersistenceCalledOutsideExample < Base
+          MSG = 'Persistence called outside of example.'
+
+          def on_send(node)
+            return if inside_example_scope?(node) ||
+              inside_method_definition?(node) ||
+              inside_proc_or_lambda?(node) ||
+              allowed_method?(node)
+            return unless inside_describe_block?(node)
+            return unless persistent_call?(node)
```

Comment:
I haven’t done benchmarks, but it looks like changing the order of these early `return`s would be good for performance. I think `persistent_call?` and `allowed_method?` are significantly faster than traversing node ancestors. How about e.g.
```rb
def on_send(node)
  return if allowed_method?
  return unless persistent_call?
  return unless inside_describe_block?
  return if lazily_executed? # as per @pirj's suggestion
  
  add_offense(node)
end
```

Reply:
Speaking on the performance, it might be better to start from an example group, and check recursively block/send nodes, breaking on valid scopes, instead of triggering the code on each send node.